### PR TITLE
barrier implementation using semaphores

### PR DESCRIPTION
**Pull request by tatehaga**
Pretty basic barrier implementation using semaphores, needed a good way to monitor threads so hopefully others can use it in the future.  Just basic barrierInit and barrierWait functions, no attributes.

### CODE REVIEW

File: nx/include/switch/kernel/barrier.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: Add comment block header like other .h.

File: nx/include/switch/kernel/barrier.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+#include "semaphore.h"
+
+
+// barrier structure
+typedef struct barrier{
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: Barrier

File: nx/include/switch/kernel/semaphore.h [hunk 3]
```diff
@@ -42,3 +42,5 @@ void semaphoreWait(Semaphore *s);
  * @return true if no wait and successful lock, false otherwise.
  */
 bool semaphoreTryWait(Semaphore *s);
+
```
Comment (yellows8) [inline nx/include/switch/kernel/semaphore.h]: revert

File: nx/source/kernel/barrier.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    my_barrier->count = 0;
+    my_barrier->thread_total = thread_count;
+    semaphoreInit(&my_barrier->throttle, 0);
+    semaphoreInit(&my_barrier->lock, 1);
+    semaphoreInit(&my_barrier->thread_wait, 0);
+}
+
+void barrierWait(barrier *b){
+    semaphoreWait(&b->lock);
+    if(b->count < b->thread_total){
```
Comment (yellows8) [inline nx/source/kernel/barrier.c]: Add spaces before/after (), likewise below.

File: nx/source/kernel/barrier.c [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#include "kernel/barrier.h"
+
+void barrierInit(barrier *my_barrier, u64 thread_count){
```
Comment (yellows8) [inline nx/source/kernel/barrier.c]: ) {, also below.

File: nx/include/switch/kernel/barrier.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+#include "semaphore.h"
+
```
Comment (fincs) [inline nx/include/switch/kernel/barrier.h]: Remove extra newline

File: nx/include/switch/kernel/barrier.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/**
+ * @file barrier.h
+ * @brief Multi-threading Barrier
+ * @author tatehaga
+ * @copyright libnx Authors
+ */
+#pragma once
+#include "semaphore.h"
+
+// barrier structure
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: /// Barrier

File: nx/include/switch/kernel/barrier.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+typedef struct Barrier{
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: Space missing before {.

File: nx/include/switch/kernel/barrier.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    u64 thread_total;
+    Semaphore throttle;
+    Semaphore lock;
+    Semaphore thread_wait;
+} Barrier;
+
+/**
+ * @brief Initializes a barrier and the number of threads to wait on.
+ * @param b Barrier object.
+ * @param thread_count initial value for the number of threads the barrier must wait for.
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: Initial

File: nx/include/switch/kernel/barrier.h [hunk 10]
```diff
@@ -7,19 +7,19 @@
...
-// barrier structure
-typedef struct Barrier{
-    u64 count;
-    u64 thread_total;
-    Semaphore throttle;
-    Semaphore lock;
-    Semaphore thread_wait;
+/// Barrier structure
+typedef struct Barrier {
+    u64 count; ///< Number of threads to reach the barrier
```
Comment (fincs) [inline nx/include/switch/kernel/barrier.h]: Align these comments.

File: nx/include/switch/kernel/barrier.h [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * @brief Multi-threading Barrier
+ * @author tatehaga
+ * @copyright libnx Authors
+ */
+#pragma once
+#include "semaphore.h"
+
+/// Barrier structure
+typedef struct Barrier {
+    u64 count;             ///< Number of threads to reach the barrier
```
Comment (yellows8) [inline nx/include/switch/kernel/barrier.h]: periods

**Pull request merged**